tools/ctl/ctl.c

```diff
@@ -42,6 +42,9 @@ struct ctl_data {
 	uint32_t type;
 	/* set or get control value */
 	bool set;
+	/* print ABI header */
+	bool print_abi_header;
```

Comment:
Hi @singalsu we can already dump/print ABI header, please see the help:
```
Usage:	 ./sof-ctl [-D <device>] [-c <control name>] [-s <data>]
	 ./sof-ctl [-D <device>] [-n <control id>] [-s <data>]
	 ./sof-ctl -h

Where:
 -D device name (default is hw:0)
 -c control name e.g. numid=22,name=\"EQIIR1.0 EQIIR\""
 -n control id e.g. 22
 -s set data using ASCII CSV input file
 -b set/get control in binary mode(e.g. for set, use binary input file, for get, dump out in hex format)
 -r no abi header for the input file, or not dumping abi header for get.
 -o specify the output file.
 -t specify the component specified type.
```
such as:
```
root@test-cml:/home/test# ./sof-ctl -n 15
Control size is 300000.
Retrieving configuration for device hw:0 control numid=15.
Success.
hdr: magic 0x00464f53
hdr: type 1
hdr: size 64 bytes
hdr: abi 3:10:0
4607827,1,64,50372608,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
root@test-cml:/home/test# 
```

Reply:
This is not the same thing. The usage for "-p" happens at audio tuning computer, e.g. computer where topology is defined. It needs to work with without (ssh) connection to a SOF device running a suitable topology. E.g. if I wan't a header for 128 bytes long ext bytes control:

```
$ sof-ctl -p 128
hdr: magic 0x00464f53
hdr: type 0
hdr: size 128 bytes
hdr: abi 3:9:0
4607827,0,128,50368512,0,0,0,0
```
You can see from the EQ tool commit how the command is used. This is generic and this 32 bytes header (currently) is not specific for EQs.

Reply:
I didn't want to change the code more than needed so the EQ tool uses the binary file output via "-b -f file.bin". The ASCII dump with the "hdr: " infos would be more complex to parse. So I just fread() as unsigned bytes the binary header blob. But a tool could as well parse this ASCII output if preferred.

Reply:
@singalsu so do you mean on host Matlab, you need to send blob with different abi header(128B comparing with 32B)? do you have the struct definition for those 128 bytes? By the way, the hdr->size is for the payload, not for the header size.

Reply:
if you want to generate different header, we need define how to generate it, if you only want to dump the first 128bytes of the tlv kcontrol, we can just dump all of it and select the first 128B?

Reply:
When we create hex blobs like SOF/tools/topology/eq*.m4 we need to know the generic binary ABI header that has changed every now and then when ABI is bumped for major version. That's currently the first 32 bytes. From EQ tools point-of-view the following EQ specific configuration request has been stable. This part is variable length (the longer the more challenging the acoustics is), so I needed the size parameter for "-p size" (with ABI header size excluded as you say). The tool can't have knowledge of length of this header and where the magic value, version, and size related bytes data need to be placed.

Therefore the EQ design tool that today can't parse C headers needs an utility to report the generic header part. The above example "sof-ctl -n 15" output is not of any help in this task. We don't know what control # the EQ will be assigned and we don't usually have SOF on the computer when Matlab/Octave runs. Also we should not depend on ssh connection to a device when some simple EQ shape like generic 50 Hz high-pass is designed. If we have an up-to-date sof-ctl utility in the computer where the EQ tool run it is able to give the correct blob header. If the created blob fails the ABI check the tuning engineer should update the sof-ctl executable that carries this information from SOF build.

Do you see problems with this idea?

Reply:
@keyonjie this option intends to help external tools to prepend the correct ABI data for tuning.

Reply:
We discussed this this today this with @keyonjie. He proposed to rename new switch to "-g" for "generate header use case". It's the third use after set/get data. Is this OK change @lgirdwood ?

Reply:
@singalsu @keyonjie yes, as long as switch is unique and you guys agree on use case.

Reply:
I've now renamed the switch to -g, the features are now all implemented.